### ISSUE TITLE
protectedFields user white-listing  not working for custom classes

### ISSUE THREAD

**Issue by Dobbias** [parse-community/parse-server#5884]
### Issue Description

When using `protectedFields` on a custom class and adding a userId to the object ACL (read) in the custom class, the ACL is ignored and the protected fields are not returned when the specified user is logged in and requests the document. Using roles works with ACL and `protectedFields`.

The problem can be traced back to the `addProtectedFields` function in [DatabaseController.js](https://github.com/parse-community/parse-server/blob/master/src/Controllers/DatabaseController.js):

The following check in the function: 
`    if (
      Object.keys(query).length === 0 &&
      auth &&
      auth.user &&
      aclGroup.indexOf(auth.user.id) > -1
    )
      return null;`

does support the behavior of white-listing users for the `protectedFields` but since `Object.keys(query).length === 0 ` almost always is larger than 0 for queries the condition is almost never true, besides the user having permission to read the `protectedFields`.

### Steps to reproduce

1. Create a custom class
2. Add protected fields
3. Create an object in the custom class
4. Add a user to the object ACL with read permission
5. Request the object as the user added to the ACL
6. Protected fields are not returned 
 
### Expected Results

The protected fields are returned since the user has the read ACL permission.

### Actual Outcome

The protected fields are not returned.

### Environment Setup

- **Server**
  - parse-server version: 3.7.2
  - Operating System: Arch Linux
  - Hardware: [FILL THIS OUT]
  - Localhost or remote server?: localhost

- **Database**
  - MongoDB version: 4.0.11
  - Storage engine:  [FILL THIS OUT]
  - Hardware:        [FILL THIS OUT]
  - Localhost or remote server?: mongoDB Atlas

**Comment by davimacedo** [comment]
@Dobbias thanks for reporting and investigating the issue. Would you be willed to send a PR with a test case and your proposed solution?

**Comment by Dobbias** [comment]
@davimacedo I'll work on it next week, if I got some spare time.

**Comment by Dobbias** [comment]
@davimacedo I investigated the code further and this actually is no bug. I referenced the code above since I had no understanding of how ACLs are applied. After having a closer look at the codebase, I would say this is a missing feature. Currently the `protectedFields` only work with roles on custom objects or on the User object using the user id (by checking the object id instead of ACL). 

It would be useful if we could define the `protectedFields` on custom classes and additionally to managing the field access by roles, manage the access by adding userIds to the ACL.

This would require one change worthy of discussion:
Currently the fields to remove (`protectedFields`) are determined before querying the objects from the database. This has the advantage of being able to only select the required fields from the database (minimizing traffic between the db and the server). If we added support for user-whitelisting of `protectedFields` through ACLs, the filtering of the fields has to be done after retrieving the object from the database, since only then the object ACLs are available to parse-server.
